File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/AppIntroModels.kt
package com.pillyliu.pinprofandroid

import androidx.annotation.DrawableRes
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import com.pillyliu.pinprofandroid.ui.PinballThemeTokens

internal enum class AppIntroProfessorSide {
    Left,
    Right,
}

internal enum class AppIntroCard(
    val title: String?,
    val subtitle: String?,
    val quote: String,
    val highlightedQuotePhrase: String?,
    val accent: Color,
    @get:DrawableRes val artworkResId: Int,
    val artworkAspectRatio: Float,
    val showsProfessorSpotlight: Boolean,
    val professorSide: AppIntroProfessorSide,
) {
    Welcome(
        title = null,
        subtitle = null,
        quote = "Welcome to PinProf, a pinball study app. Go from pinball novice to pinball wizard in no time!",
        highlightedQuotePhrase = "PinProf",
        accent = AppIntroTheme.glow,
        artworkResId = R.drawable.intro_launch_logo,
        artworkAspectRatio = 1f,
        showsProfessorSpotlight = false,
        professorSide = AppIntroProfessorSide.Left,
    ),
    League(
        title = "League",
        subtitle = "Lansing Pinball League stats",
        quote = "Among peers, statistics reveal true standing.",
        highlightedQuotePhrase = null,
        accent = Color(0xFFC3C978),
        artworkResId = R.drawable.intro_league_screenshot,
        artworkAspectRatio = 1206f / 1809f,
        showsProfessorSpotlight = true,
        professorSide = AppIntroProfessorSide.Left,
    ),
    Library(
        title = "Library",
        subtitle = "Rulesheets, playfields, tutorials",
        quote = "Attend closely; mastery follows diligence.",
        highlightedQuotePhrase = null,
        accent = Color(0xFF8FDBC7),
        artworkResId = R.drawable.intro_library_screenshot,
        artworkAspectRatio = 1206f / 1809f,
        showsProfessorSpotlight = true,
        professorSide = AppIntroProfessorSide.Right,
    ),
    Practice(
        title = "Practice",
        subtitle = "Track practice, trends, progress",
        quote = "A careful record reveals true progress.",
        highlightedQuotePhrase = null,
        accent = Color(0xFFFFDB66),
        artworkResId = R.drawable.intro_practice_screenshot,
        artworkAspectRatio = 1206f / 1809f,
        showsProfessorSpotlight = true,
        professorSide = AppIntroProfessorSide.Left,
    ),
    GameRoom(
        title = "GameRoom",
        subtitle = "Organize machines and upkeep",
        quote = "Order and care are marks of excellence.",
        highlightedQuotePhrase = null,
        accent = Color(0xFFF5C75C),
        artworkResId = R.drawable.intro_gameroom_screenshot,
        artworkAspectRatio = 1206f / 1809f,
        showsProfessorSpotlight = true,
        professorSide = AppIntroProfessorSide.Right,
    ),
    Settings(
        title = "Settings",
        subtitle = "Sources, venues, tournaments, data",
        quote = "A well-curated library reflects discernment.",
        highlightedQuotePhrase = null,
        accent = Color(0xFFB8E5C2),
        artworkResId = R.drawable.intro_settings_screenshot,
        artworkAspectRatio = 1206f / 1809f,
        showsProfessorSpotlight = true,
        professorSide = AppIntroProfessorSide.Left,
    ),
}

internal object AppIntroTheme {
    val tint = Color(0xFF1F5742)
    val glow = Color(0xFFA3E0BD)
    val text = Color.White.copy(alpha = 0.96f)
    val secondaryText = Color.White.copy(alpha = 0.84f)
}

internal object AppIntroTypography {
    val title = FontFamily(
        Font(R.font.bodoni_moda_variable, weight = FontWeight.Bold),
    )
    val subtitle = FontFamily(
        Font(R.font.cormorant_garamond_variable, weight = FontWeight.SemiBold),
    )
    val quote = FontFamily(
        Font(
            R.font.libre_baskerville_italic_variable,
            weight = FontWeight.SemiBold,
            style = FontStyle.Italic,
        ),
        Font(
            R.font.libre_baskerville_italic_variable,
            weight = FontWeight.Bold,
            style = FontStyle.Italic,
        ),
    )
}

@Composable
internal fun AppIntroCard.resolvedAccent(): Color {
    return when (this) {
        AppIntroCard.League -> PinballThemeTokens.colors.statsMeanMedian
        else -> accent
    }
}

File: Pinball App 2/Pinball App 2/app/AppIntroModels.swift
import SwiftUI
import UIKit

enum AppIntroProfessorSide {
    case left
    case right
}

enum AppIntroCard: Int, CaseIterable, Identifiable {
    case welcome
    case league
    case library
    case practice
    case gameroom
    case settings

    var id: Int { rawValue }

    var title: String? {
        switch self {
        case .welcome:
            return nil
        case .league:
            return "League"
        case .library:
            return "Library"
        case .practice:
            return "Practice"
        case .gameroom:
            return "GameRoom"
        case .settings:
            return "Settings"
        }
    }

    var subtitle: String? {
        switch self {
        case .welcome:
            return nil
        case .league:
            return "Lansing Pinball League stats"
        case .library:
            return "Rulesheets, playfields, tutorials"
        case .practice:
            return "Track practice, trends, progress"
        case .gameroom:
            return "Organize machines and upkeep"
        case .settings:
            return "Sources, venues, tournaments, data"
        }
    }

    var quote: String {
        switch self {
        case .welcome:
            return "Welcome to PinProf, a pinball study app. Go from pinball novice to pinball wizard in no time!"
        case .league:
            return "Among peers, statistics reveal true standing."
        case .library:
            return "Attend closely; mastery follows diligence."
        case .practice:
            return "A careful record reveals true progress."
        case .gameroom:
            return "Order and care are marks of excellence."
        case .settings:
            return "A well-curated library reflects discernment."
        }
    }

    var highlightedQuotePhrase: String? {
        switch self {
        case .welcome:
            return "PinProf"
        case .league, .library, .practice, .gameroom, .settings:
            return nil
        }
    }

    var accent: Color {
        switch self {
        case .welcome:
            return AppIntroTheme.glow
        case .league:
            return AppTheme.statsMeanMedian
        case .library:
            return Color(red: 0.56, green: 0.86, blue: 0.78)
        case .practice:
            return Color(red: 1.00, green: 0.86, blue: 0.40)
        case .gameroom:
            return Color(red: 0.96, green: 0.78, blue: 0.36)
        case .settings:
            return Color(red: 0.72, green: 0.90, blue: 0.76)
        }
    }

    var bundledArtworkFileName: String {
        switch self {
        case .welcome:
            return "launch-logo.webp"
        case .league:
            return "league-screenshot.webp"
        case .library:
            return "library-screenshot.webp"
        case .practice:
            return "practice-screenshot.webp"
        case .gameroom:
            return "gameroom-screenshot.webp"
        case .settings:
            return "settings-screenshot.webp"
        }
    }

    var artworkAspectRatio: CGFloat {
        switch self {
        case .welcome:
            return 1.0
        case .league, .library, .practice, .gameroom, .settings:
            return 1206.0 / 1809.0
        }
    }

    var showsProfessorSpotlight: Bool {
        self != .welcome
    }

    var professorSide: AppIntroProfessorSide {
        switch self {
        case .welcome, .league, .practice, .settings:
            return .left
        case .library, .gameroom:
            return .right
        }
    }
}

enum AppIntroTheme {
    static let tint = Color(red: 0.12, green: 0.34, blue: 0.26)
    static let glow = Color(red: 0.64, green: 0.88, blue: 0.74)
    static let text = Color.white.opacity(0.96)
    static let secondaryText = Color.white.opacity(0.84)
}

extension Font {
    static func appIntroTitle(size: CGFloat) -> Font {
        let preferredNames = [
            "Didot-Bold",
            "BodoniSvtyTwoITCTT-Bold",
            "AvenirNextCondensed-Heavy"
        ]

        if let fontName = preferredNames.first(where: { UIFont(name: $0, size: size) != nil }) {
            return .custom(fontName, size: size, relativeTo: .title3)
        }

        return .system(size: size, weight: .bold, design: .rounded)
    }

    static func appIntroSubtitle(size: CGFloat) -> Font {
        let preferredNames = [
            "Optima-Regular",
            "GillSans-SemiBold",
            "AvenirNext-DemiBold"
        ]

        if let fontName = preferredNames.first(where: { UIFont(name: $0, size: size) != nil }) {
            return .custom(fontName, size: size, relativeTo: .subheadline)
        }

        return .system(size: size, weight: .semibold, design: .rounded)
    }

    static func appIntroQuote(size: CGFloat) -> Font {
        let preferredNames = [
            "Baskerville-SemiBoldItalic",
            "Baskerville-Italic",
            "TimesNewRomanPS-ItalicMT"
        ]

        if let fontName = preferredNames.first(where: { UIFont(name: $0, size: size) != nil }) {
            return .custom(fontName, size: size, relativeTo: .body)
        }

        return .system(size: size, weight: .semibold, design: .serif).italic()
    }

    static func appIntroQuoteHighlighted(size: CGFloat) -> Font {
        let preferredNames = [
            "Baskerville-BoldItalic",
            "Baskerville-SemiBoldItalic",
            "TimesNewRomanPS-BoldItalicMT"
        ]

        if let fontName = preferredNames.first(where: { UIFont(name: $0, size: size) != nil }) {
            return .custom(fontName, size: size, relativeTo: .body)
        }

        return .system(size: size, weight: .bold, design: .serif).italic()
    }
}
